### PR TITLE
Sstables: limit snapshot dirs

### LINKED ISSUE #5324: snapshot only sstables in base directory
We've hit a case where snapshot failed since an sstable in staging got removed while taking its snapshot. We need to snapshot only sstables located in the base table directory, not in "staging" or "uploads".

### PR DESCRIPTION
**Pull request by bhalevy**
Fixes #5324

We've hit a case where snapshot failed since an sstable in staging got removed while taking its snapshot. We need to snapshot only sstables located in the base table directory, not in "staging" or "uploads".

The patchset introduces an enum class sstable::location_types that indicates where the sstable is in the data hierarchy: base, staging, upload, or snapshots.

The sstable location is determined when the sstable object is constructed using the sstable directory using a regular expression shared with `make_descriptor`.  `sstable::get_location()` retrieves the location.

The sstable location is used by `table::snapshot` to filter `_sstables->all()` and ignore any sstable that is not in the base directory.

A unit test for sstable location has been added to `sstable_test`.

### CODE REVIEW

File: sstables/entry_descriptor.hh [hunk 1]
```diff
@@ -21,6 +21,8 @@
 
 #pragma once
```
Comment (avikivity) [inline sstables/entry_descriptor.hh]: What's the motivation? You're declaring static file/dir types, but I can't guess why.

File: table.cc [hunk 2]
```diff
@@ -1720,7 +1720,10 @@ seal_snapshot(sstring jsondir) {
 future<> table::snapshot(sstring name) {
```
Comment (avikivity) [inline table.cc]: Fixes ????
Explanation?

**Comment by avikivity** [comment]
looks okay, needs documentary improvement

**Comment by nyh** [comment]
@bhalevy I'm trying to understand your statement that "we need to snapshot only sstables located in the base table directory, not in "staging" or "uploads"."
Please correct me if I'm wrong (it's been a while since I looked at that stuff), but I think that sstables in staging/ are already readable by a client, aren't they? In this case, why should they not be snapshotted as well?

File: sstables/entry_descriptor.hh (hunk 1 above)
Comment (bhalevy) [reply sstables/entry_descriptor.hh]: We're practically using `pragma once` as an include guard.
Isn't that by convention?

File: table.cc (hunk 2 above)
Comment (bhalevy) [reply table.cc]: The motivation is described in the pull request message.
I can add a comment here as well.

**Comment by bhalevy** [comment]
> @bhalevy I'm trying to understand your statement that "we need to snapshot only sstables located in the base table directory, not in "staging" or "uploads"."
> Please correct me if I'm wrong (it's been a while since I looked at that stuff), but I think that sstables in staging/ are already readable by a client, aren't they? In this case, why should they not be snapshotted as well?

I was under the impression from @slivne's comment that we should not take snapshots in staging directory, but after discussion and re-reading the code I think you're right and these are readable and need to be part of the snapshot.  That said, [apparently](https://github.com/scylladb/scylla-enterprise/issues/1146) we need to do that carefully.

**Comment by bhalevy** [comment]
Retracting this PR in favor of proper synchronization of snapshot and view_update_generator.

File: sstables/sstables.cc [hunk 3]
```diff
@@ -2697,7 +2720,7 @@ void sstable::move_to_new_dir_in_thread(sstring new_dir, int64_t new_generation)
 
 std::regex entry_descriptor::la_mc_filename_pattern("(la|mc)-(\\d+)-(\\w+)-(.*)");
 std::regex entry_descriptor::ka_filename_pattern("(\\w+)-(\\w+)-ka-(\\d+)-(.*)");
-std::regex entry_descriptor::sst_dir_pattern(".*/([^/]*)/([^/]+)-[\\da-fA-F]+(?:/staging|/upload|/snapshots/[^/]+)?/?");
+std::regex entry_descriptor::sst_dir_pattern(".*/([^/]*)/([^/]+)-[\\da-fA-F]+(/(staging|upload|snapshots)(/[^/]+)?)?/*");
```
Comment (denesb) [inline sstables/sstables.cc]: I don't see the relaxing, the `staging|upload|snapshot` part is now mandatory. Furthermore there is now an unmatched `(`. Also, this pattern now permits any amount of trailing `/`.

File: sstables/types.hh [hunk 4]
```diff
@@ -49,6 +49,14 @@ static inline bytes_view to_bytes_view(const temporary_buffer<char>& b) {
 
 namespace sstables {
 
+enum class sstable_location_types {
```
Comment (denesb) [inline sstables/types.hh]: Why the plural?

**Pull request closed without merging**